pkg/cri/resource-manager/policy/builtin/memtier/pools.go

```diff
@@ -27,108 +27,237 @@ import (
+		//
+		//   The PMEM memory of (omitted) PMEM-only nodes is assigned
+		//   to one of the closest normal (DRAM) NUMA nodes. This right
+		//   assignment has already been calculated by assignPMEMNodes().
+		//   However, making the corresponding assignment in the pool
+		//   tree is a bit more involved as the DRAM node where a PMEM
+		//   node has been assigned to might have gotten omitted from the
+		//   tree if it ended up being a lone child. We use the recorded
+		//   per NUMA node surrogates to find both if and where resources
+		//   of omitted DRAM NUMA nodes need to get assigned to, and also
+		//   where PMEM NUMA node resources need to get assigned to.
+		//
+
+		if n.IsLeafNode() {
+			for numaID, surrogate := range numaSurrogates {
+				if n.IsSameNode(surrogate) {
+					// Notes:
+					//   Only assign DRAM NUMA node resources to genuine
+					//   surrogates of omitted NUMA nodes, never to the
+					//   pool NUMA node itself (since then it has not
+					//   been omitted and resources for it were properly
+					//   discovered above). There is no need to test the
+					//   pool nodes NUMA ID for this check, because a NUMA
+					//   pool node is only ever set as (a fake) surrogate
+					//   for itself.
+					if n.Kind() != NumaNode {
+						n.AssignNUMANodes([]system.ID{numaID})
+					}
+					if pmem, ok := assigned[numaID]; ok {
+						n.AssignNUMANodes(pmem)
+						delete(assigned, numaID)
+					}
 				}
 			}
 		}
+
+		return nil
+	})
+
+	// make sure all PMEM nodes got assigned
+	if len(assigned) > 0 {
+		for numaID, pmem := range assigned {
+			log.Error("failed to assign PMEM NUMA nodes #%s (to NUMA node/surrogate #%d)",
+				system.NewIDSet(pmem...), numaID)
+		}
+		log.Fatal("internal error: unassigned PMEM NUMA nodes remaining")
 	}
 
-	pmemNodes := map[system.ID]system.Node{}
-	dramNodes := map[system.ID]system.Node{}
+	p.root.Dump("<pool-setup>")
 
-	// create nodes for NUMA nodes
-	if nodeCnt > 0 {
-		for _, id := range p.sys.NodeIDs() {
-			var parent Node
+	return nil
+}
 
-			sysnode := p.sys.Node(id)
-			switch sysnode.GetMemoryType() {
-			case system.MemoryTypePMEM:
-				pmemNodes[id] = sysnode
-				continue // don't create pool nodes for CPU-less NUMA nodes
-			case system.MemoryTypeDRAM:
-				dramNodes[id] = sysnode
+// parentNumaNodeCountWithCPUs returns the number of CPU-ful NUMA nodes in the parent die/socket.
+func (p *policy) parentNumaNodeCountWithCPUs(numaNode system.Node) int {
+	socketID := numaNode.PackageID()
+	socket := p.sys.Package(socketID)
+	dieIDs := socket.DieIDs()
+	if len(dieIDs) > 1 {
+		// check if we have more than one cpu-ful NUMA node in the same die as the given NUMA node
+		for _, dieID := range dieIDs {
+			found := false
+			count := 0
+			for _, nodeID := range socket.DieNodeIDs(dieID) {
+				node := p.sys.Node(nodeID)
+				if !node.CPUSet().IsEmpty() {
+					count++
+				}
+				if nodeID == numaNode.ID() {
+					found = true
+				}
```

Comment:
Stupid question to understand the code better: aren't we able to get the die of the node directly?

Reply:
No, not with the current `pkg/sysfs` code. That only tracks 'first-class' topology objects, those with a system-wide unique ID, explicitly. Die IDs are only unique within their socket. But that shouldn't prevent us from storing the die ID in the node anyway. I'll change that. It makes more sense, since we store the package ID there, too..